src/expr.rs

```diff
@@ -688,6 +688,13 @@ fn rewrite_closure_expr(
     if classify::expr_requires_semi_to_be_stmt(left_most_sub_expr(expr)) {
         rewrite = and_one_line(rewrite);
     }
+    rewrite = rewrite.and_then(|rw| {
+        if !context.config.overflow_closures() && rw.contains('\n') {
```

Comment:
Look at how `rewrite_closure_expr` is used:

https://github.com/rust-lang-nursery/rustfmt/blob/6fb17c41301e5a98748882cc6ad6a627def8eb80/src/expr.rs#L629-L648

https://github.com/rust-lang-nursery/rustfmt/blob/6fb17c41301e5a98748882cc6ad6a627def8eb80/src/expr.rs#L650-L653

It's called to attempt to format the closure body expression without wrapping it in a block. If it fails and returns `None`, rustfmt will fall back to `rewrite_closure_block` and format the expression with a block surrounding it. `rewrite_closure_block` will perform any formatting that's still appropriate for the expression in that context.